### ISSUE TITLE
CalculateDesiredIncludesAndForwardDeclares ends up duplicating OneIncludeOrForwardDeclareLine

### ISSUE THREAD

**Issue by dpunset** [include-what-you-use/include-what-you-use#215]
Hello,

This is related to #213, but it's not a case sensitive issue anymore.

I was having a case in which IWYU was suggesting me to remove an include that was needed for sure. I debugged the case, and it's a bit complex but it keeps showing flaws on this function.

With some structure like this:

c:\mycode\libA\somefolder\ClassA.cpp
c:\mycode\libA\somefolder\ClassA.h
c:\mycode\libA\somefolder\Enums.h

I have ClassA in ClassA.cpp and ClassA.h. Because of a series of indirect includes, ClassA.h ends up being included by the precompiled header, therefore marked as a prefix include. ClassA.cpp is the file I use to test with IWYU.

When IWYU analyze this TU, it tells me to remove Enums.h in ClassA.h, which is needed to declare a member of the class as an enum. So I can't do that.

The reason why IWYU tells me to do so is a bit complex. I noticed that the desired flag was being removed from Enums.h (in CleanupPrefixHeaderIncludes), and it was cleared because the is_present flag was not set. Since is_present should be true because Enums.h is directly written in the ClassA.h I checked deeper, and I noticed that again Enums.h was appearing twice in the lines list.

Enums.h gets added a second time in CalculateDesiredIncludesAndForwardDeclares because the suggested header for the use being processed is different than the line already added in the list:
#include "libA\somefolder\Enums.h"
#include "somefolder\Enums.h"

The search paths are -I c:\mycode\ -I c:\mycode\libA

In ClassA.h, Enums.h is referenced like this:

#include "libA\somefolder\Enums.h"

Which matches with the first time the usage is found in the lines array.
However, for the second case, the suggested header is taken from the public_headers, which are computed when calling SetPublicHeaders with GetCandidateHeadersForFilepathIncludedFrom, which uses the search paths to convert things to quoted includes. In the search paths list c:\mycode\libA is found before c:\mycode\, which results in stripping the libA part, which we always put in our includes (except for precompiled headers, that's why it's there). In the end, the suggested header appears like this:

#include "somefolder\Enums.h"

Finally, since this is a line that gets added at the end, it's not flagged as present, and CleanupPrefixHeaderIncludes decides to clear the desired flag from Enums.h

Don't you think we should never rely on the "suggested_header" for anything? Wouldn't it be better to just rely on the file referenced itself? (clang::FileEntry) It doesn't seem possible to get a consistent naming since sometimes things are included manually, others from other headers with different paths, names get stripped and issues like this can happen.

**Comment by dpunset** [comment]
I tried a fix locally that seems to do it, but I am not sure if it will break any tests.
These are the changes:

iwyu_output.h
I added to class OneIncludeOrForwardDeclareLine
  bool matches(const clang::FileEntry* file) const {
      return IsIncludeLine() && (included_file_ == file);
  }

iwyu_output.cc
Inside CalculateDesiredIncludesAndForwardDeclares
if (use.is_full_use()) {
      CHECK_(use.has_suggested_header() && "Full uses should have #includes");
-     if (!Contains(*lines, use.suggested_header())) { // must be added
+    if (!Contains(*lines, use.decl_file())) { // must be added
        lines->push_back(OneIncludeOrForwardDeclareLine(
            use.decl_file(), use.suggested_header(), -1));
      }

  typedef multimap<string, OneIncludeOrForwardDeclareLine*> IncludeMap;
+  typedef multimap<const clang::FileEntry*, OneIncludeOrForwardDeclareLine*> IncludeMapFile;
  typedef multimap<const NamedDecl*, OneIncludeOrForwardDeclareLine*>

  IncludeMap include_map;
+  IncludeMapFile include_map_file;
  FwdDeclMap fwd_decl_map;

    if (line.IsIncludeLine()) {
        include_map.insert(std::make_pair(line.quoted_include(), &line));
+        include_map_file.insert(std::make_pair(line.included_file(), &line));
    }

+      if (range.first != range.second) {
          for (auto it = range.first; it != range.second; ++it) {
              it->second->set_desired();
              it->second->AddSymbolUse(use.short_symbol_name());
          }
+      }
+      else {
+          auto range = include_map_file.equal_range(use.decl_file());
+          for (auto it = range.first; it != range.second; ++it) {
+              it->second->set_desired();
+              it->second->AddSymbolUse(use.short_symbol_name());
+          }
+      }

Sorry for not doing a normal dif but I also have the changes for the case sensitive issue I mentioned in another post.

This does fix my current issue but I don't think it's fixing everything. I don't know the full scope, but it feels that all the tests for comparing string headers should be comparing file entries instead. What do you think?

**Comment by kimgr** [comment]
1) Build IWYU
2) Add the directory where include-what-you-use.exe got built to your PATH environment variable
3) Run `python run_iwyu_tests.py` in the IWYU source directory (requires Python, of course)

You should see three failures on Windows (badinc, iterator, stl_container_provides_allocator).

I'll try to take a look at the problem and your changes soon...

**Comment by dpunset** [comment]
It seems like I don't pass the tests unfortunately. I added some asserts to
make sure that the lines have a included_file() and some don't. For
example, #include <new>. I don't know why it tries to add that sort of
thing though. The assert always triggers with "new" when trying to build
the map (include_map).
This is the code:

// We'll want to be able to map from an include or fwd-declare to the
  // lines where we found them.  There can be multiple includes or
fwd-declares
  // providing the same symbol, so use multimaps for these reverse lookups.
  typedef multimap<const clang::FileEntry*,
OneIncludeOrForwardDeclareLine*> IncludeMapFile;
  typedef multimap<const NamedDecl*, OneIncludeOrForwardDeclareLine*>
      FwdDeclMap;

  IncludeMapFile include_map_file;
  FwdDeclMap fwd_decl_map;

  for (OneIncludeOrForwardDeclareLine& line : *lines) {
    if (line.IsIncludeLine()) {
      CHECK_(line.included_file() != nullptr && "Included file not
present");
      include_map_file.insert(std::make_pair(line.included_file(), &line));
    }
    else
      fwd_decl_map.insert(std::make_pair(line.fwd_decl(), &line));
  }

**Comment by dpunset** [comment]
Oh and by the way I had to modify iwyu_test_util.py to support paths with spaces towards the iwyu executable.

At line 406:
-cmd = '%s %s %s %s' % (
+cmd = '"%s" %s %s %s' % (

**Comment by dpunset** [comment]
Hey @kimgr . I wanted to compare the tests done from a clean changelist of IWYU and mine. I got your pull about the macro locations as is and I ended up having 84 failures. I tested with the master changelist and I still got 84 failures. You said I should only get 3 failures, right?

**Comment by kimgr** [comment]
Yes, three are expected on Windows, we haven't been able to sort them out.

So, what seems to be the cause of the failures? Sounds like it can't find IWYU or something.

**Comment by dpunset** [comment]
Oh it does, I copied into one of the directories in the path, and I made sure that it was found and used. It's just that the tests fail, I have a dump of the tests, but unfortunatelly I can't upload it here :( I will send it to you by email if you don't mind.

**Comment by kimgr** [comment]
Sure. I'm guessing it's huge, but go ahead.

**Comment by vsapsai** [comment]
@dpunset, answering your early question. Yes, it is more correct to compare files than file paths. Different string manipulations to improve path handling aren't really solving the problem. For issue #213 I am checking exactly what do we need to change to compare files, not file paths.

**Comment by kimgr** [comment]
> Oh and by the way I had to modify iwyu_test_util.py to support paths with spaces towards the iwyu
> executable.
>
> At line 406:
> -cmd = '%s %s %s %s' % (
> +cmd = '"%s" %s %s %s' % (

I'd like to clean this up more generally in iwyu_test_util.py, I'll open a new issue for it.

**Comment by kimgr** [comment]
> For example, #include <new>. I don't know why it tries to add that sort of
> thing though. The assert always triggers with "new" when trying to build
> the map (include_map).

`<new>` is special somehow, I can't remember exactly. I think there's some wording in the standard that says it needs to be included if you're using placement new. In `VisitCXXNewExpr`, IWYU conjures up an include dependency to `<new>`, but it doesn't have enough metadata to construct a proper `FileEntry*`.

If you can work around that you should be on your way to something that might work.
